editor/plugins/mesh_instance_3d_editor_plugin.cpp

```diff
@@ -103,31 +107,249 @@ Vector<Ref<Shape3D>> MeshInstance3DEditor::create_shape_from_mesh(Ref<Mesh> p_me
+				cylinder_shape->set_radius((cylinder_mesh->get_top_radius() + cylinder_mesh->get_bottom_radius()) / 2.0);
+				shapes.push_back(cylinder_shape);
+			}
+
+			const Ref<SphereMesh> sphere_mesh = p_mesh;
+			if (sphere_mesh.is_valid()) {
+				Ref<SphereShape3D> sphere_shape;
+				sphere_shape.instantiate();
+				sphere_shape->set_radius(sphere_mesh->get_radius());
+				shapes.push_back(sphere_shape);
+			}
+
+			if (p_verbose && shapes.is_empty()) {
+				err_dialog->set_text(TTR("Couldn't create a primitive collision shape."));
+				err_dialog->popup_centered();
+			}
+		} break;
+
 		default:
 			break;
 	}
 	return shapes;
 }
 
+void MeshInstance3DEditor::_shape_dialog_about_to_popup() {
+	EditorSelection *editor_selection = EditorNode::get_singleton()->get_editor_selection();
+	List<Node *> selection = editor_selection->get_selected_node_list();
+	if (selection.is_empty()) {
+		selection.push_back(node);
+	}
+
+	bool disable_primitive = true;
+	for (Node *E : selection) {
+		MeshInstance3D *instance = Object::cast_to<MeshInstance3D>(E);
+		if (!instance) {
+			continue;
+		}
+		Ref<Mesh> m = instance->get_mesh();
+		if (m.is_null()) {
+			continue;
+		}
+		if (((Ref<BoxMesh>)m).is_valid() || ((Ref<CapsuleMesh>)m).is_valid() || ((Ref<CylinderMesh>)m).is_valid() || ((Ref<SphereMesh>)m).is_valid()) {
+			disable_primitive = false;
+			break;
+		}
+	}
+	if (disable_primitive && shape_type->get_selected() == SHAPE_TYPE_PRIMITIVE) {
+		shape_type->select(SHAPE_TYPE_TRIMESH);
+	}
+
+	shape_type->set_item_disabled(shape_type->get_popup()->get_item_index(SHAPE_TYPE_PRIMITIVE), disable_primitive);
+
+	ShapeType selected_shape_type = (ShapeType)shape_type->get_selected_id();
+	bool shape_axis_visible = selected_shape_type == SHAPE_TYPE_CAPSULE || selected_shape_type == SHAPE_TYPE_CYLINDER;
+	shape_axis->set_visible(shape_axis_visible);
+	shape_axis_label->set_visible(shape_axis_visible);
+}
+
+void MeshInstance3DEditor::_shape_type_selected(int p_option) {
+	bool shape_axis_visible = (ShapeType)p_option == SHAPE_TYPE_CAPSULE || (ShapeType)p_option == SHAPE_TYPE_CYLINDER;
+	shape_axis->set_visible(shape_axis_visible);
+	shape_axis_label->set_visible(shape_axis_visible);
+}
+
 void MeshInstance3DEditor::_create_collision_shape() {
 	int placement_option = shape_placement->get_selected();
 	int shape_type_option = shape_type->get_selected();
 
 	EditorSelection *editor_selection = EditorNode::get_singleton()->get_editor_selection();
 	EditorUndoRedoManager *ur = EditorUndoRedoManager::get_singleton();
 
+	String placement_action_name = placement_option == SHAPE_PLACEMENT_SIBLING ? TTR("Create %s Collision Shape Sibling") : TTR("Create %s Static Body Child");
+
 	switch (shape_type_option) {
 		case SHAPE_TYPE_TRIMESH: {
-			ur->create_action(TTR(placement_option == SHAPE_PLACEMENT_SIBLING ? "Create Trimesh Collision Shape Sibling" : "Create Trimesh Static Body"));
+			ur->create_action(vformat(placement_action_name, TTR("Trimesh")));
```

Comment:
I believe these were incorrect for translations, so I changed how they use TTR.
It should now also be easier to extend if more placement options are added.

Reply:
Indeed, `TTR()` should not contain conditions within it (the condition should be outside).